### PR TITLE
adds instrumentation reset in harness

### PR DESCRIPTION
**Pull request by lukastruemper**
## Description
A clear and concise description of what this PR does and why.

## Related Issues

- Closes #

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [ ] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [ ] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [ ] New unit tests have been added for the changes.
- [ ] Existing unit tests pass locally.
- [ ] Edge cases and error paths are covered.

### Integration Tests
- [ ] New integration tests have been added (or existing ones updated).
- [ ] Integration tests pass locally.
- [ ] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [ ] Test coverage has not decreased as a result of this change.
- [ ] Coverage report reviewed and acceptable.
- [ ] Critical / new code paths are covered.

## How Has This Been Tested?
Describe the tests you ran to verify your changes and provide instructions to reproduce.
Include relevant details for your test configuration (OS, Python/LLVM/Docc version, hardware/target).

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: rtl/CMakeLists.txt [hunk 1]
```diff
@@ -4,7 +4,17 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 set(CMAKE_CXX_STANDARD 20)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
-add_library(daisy_rtl STATIC
+# Static by default (each artifact embeds its own instrumentation state). The
```
Comment (ramonwirsch) [inline rtl/CMakeLists.txt]: This makes llvm frontend's use of rtl / instrumentation no longer work if DAISY_RTL_SHARED is on.

Meaning, we now require 2 separate build folders, one for python-... and one for llvm builds.

I suggest we either always build both and have each front end do what it wants, or we upgrade the llvm frontend to using the dynamic library always as well.

File: CMakeLists.txt [hunk 2]
```diff
@@ -73,6 +73,16 @@ endif()
 add_subdirectory(sdfg)
 add_subdirectory(opt)
 add_subdirectory(rpc)
+
+# The python/mlir frontends need a single shared RTL instance per process so the
+# runtime bindings can reset/query it (see rtl/CMakeLists.txt). Decide the RTL
+# linkage before the rtl subdirectory is configured below.
+option(PYTHON_BUILD_FRONTEND "Build Python frontend" OFF)
```
Comment (ramonwirsch) [inline CMakeLists.txt]: this will implicitly make llvm frontend fail partially by enabling other frontends

File: c-compile/src/compile/src_file_compiler.cpp [hunk 3]
```diff
@@ -278,6 +278,9 @@ bool SrcFileCompiler::run_compile_and_link_single(const std::filesystem::path& s
 void SrcFileCompiler::add_link_args(std::stringstream& cmd) const {
     for (auto& ld : this->library_paths_) {
         cmd << "-L " << ld << " ";
+        // Embed an rpath so a shared RTL (DAISY_RTL_SHARED) resolves at runtime
```
Comment (ramonwirsch) [inline c-compile/src/compile/src_file_compiler.cpp]: This is somewhat dirty as we would also add target-specific libraries to the rpath, that we so far expect to have on LD_LIBRARY_PATH (and that users may expect to switch around after compiling)

Ok to merge, but a cleaner solution would be to have an additional flag that is set for daisy_rtl if we want to link it dynamically or for our own libs.

File: python/docc/benchmarks/rtl.py [hunk 4]
```diff
@@ -54,6 +61,140 @@
...
+    """(exact filename, glob) for the RTL on the current platform."""
+    if sys.platform == "darwin":
+        return "libdaisy_rtl.dylib", "libdaisy_rtl*.dylib"
+    if sys.platform.startswith("win"):
+        return "daisy_rtl.dll", "daisy_rtl*.dll"
+    # Match the plain name and, defensively, any soname-versioned file.
+    return "libdaisy_rtl.so", "libdaisy_rtl.so*"
+
+
+def _candidate_lib_dirs() -> List[Path]:
```
Comment (ramonwirsch) [inline python/docc/benchmarks/rtl.py:81]: C++ code discovers and tracks the various locations for its binaries and sources and uses those paths to compile and link the SO. So we do not need to try all paths from python, we could just ask the docc native code, which folders it is using.

Ok to merge, but there is a cleaner solution.

**Review by ramonwirsch — CHANGES_REQUESTED** [review]
Makes it impossible to use llvm and python- frontends from the same cmake build setup.
(at least when instrumentation is used).

We can either make the llvm frontend also support the daisy_rtl.so (by doing the same rpath) and make it default & permanently shared. Or we can setup cmake to build both and drop the option to switch which one we build. Then llvm can default to the old way and python frontends can default to the new way.

**Review by ramonwirsch — APPROVED** [review]
Resolved all concerns. Using unified shared library + rpath everywhere, with explicit management of rpath and using the paths of library in the source repository we also use for linking if we detect being run from the cmake source tree
